app/api/db/__init__.py

```diff
@@ -1,60 +1,159 @@
-    )
+class DB:
+    """Database connection manager.
+
+    This class is used to manage database connections for the Flask app.
+    It has methods for getting a new connection or session object.
+    """
+
+    _engine: sqlalchemy.engine.Engine
+
+    def __init__(self) -> None:
+        self._engine = _create_db_engine()
+
+    def init_app(self, app: Flask) -> None:
+        """Initialize the Flask app.
+
+        Add the database to the Flask app's extensions so that it can be
+        accessed by request handlers using the current app context.
+
+        see get_db
+        """
+        app.extensions[FLASK_EXTENSION_KEY] = self
 
-    if check_migrations_current:
-        have_all_migrations_run(engine)
+    def get_connection(self) -> Connection:
+        """Return a new database connection object.
 
-    engine.dispose()
+        Use the connection to execute SQL queries without using the ORM.
 
-    return session_factory
+        Usage:
+            with db.get_connection() as conn:
+                conn.execute(...)
+        """
+        return self._engine.connect()
+
+    def get_session(self) -> Session:
+        """Return a new session object.
+
+        In general, only one session object should be created per request.
+
+        If you want to automatically commit or rollback the session, use
+        the session.begin() context manager.
+        See https://docs.sqlalchemy.org/en/13/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it
+
+        Example:
+            with db.get_session() as session:
+                with session.begin():
+                    session.add(...)
+                # session is automatically committed here
+                # or rolled back if an exception is raised
+        """
+        return Session(bind=self._engine, expire_on_commit=False, autocommit=False)
+
+    def check_db_connection(self) -> None:
+        """Check that we can connect to the database and log some info about the connection."""
+        logger.info("connecting to postgres db")
+        with self.get_connection() as conn:
+            conn_info = conn.connection.dbapi_connection.info  # type: ignore
+
+            logger.info(
+                "connected to postgres db",
+                extra={
+                    "dbname": conn_info.dbname,
+                    "user": conn_info.user,
+                    "host": conn_info.host,
+                    "port": conn_info.port,
+                    "options": conn_info.options,
+                    "dsn_parameters": conn_info.dsn_parameters,
+                    "protocol_version": conn_info.protocol_version,
+                    "server_version": conn_info.server_version,
+                },
+            )
+            verify_ssl(conn_info)
+
+            # TODO add check_migrations_current to config
+            # if check_migrations_current:
+            #     have_all_migrations_run(engine)
+
+
+def init() -> DB:
```

Comment:
Thoughts on changing this to the below to avoid someone forgetting to check the connection where it makes sense to do so? I think outside tests we'd generally want to do that, right?

```py
def init(verify_connection: bool = True) -> DB:
    db = DB()
    if verify_connection:
         db.check_db_connection()

     return db
```

Reply:
Sounds good to me. I was on the fence about it, since I was generally hoping to avoid parameters in the init function. The previous init function with 3 parameters made it harder for me to reason about all the various entry points, whereas putting the the extra calls in the places that needed them made it easier for me to understand. That said I think you're right in that it actually simplifies the number of lines in `__main__` and only adds complexity to the test suite, which I'm more ok with.